frontend/src/components/pages/RegistrantExperience/Waiver/WaiverReducer.tsx

```diff
@@ -0,0 +1,78 @@
+import {
+  WaiverInterface,
+  OptionalClauseResponse,
+  RequiredClauseResponse,
+  WaiverReducerDispatch,
+  WaiverActions,
+  LoadedWaiver,
+  ClickOptionalClause,
+  FillDate,
+  FillName,
+} from "../../../../types/waiverTypes";
+import { WaiverClause } from "../../../../types/AdminTypes";
+
+const waiverReducer = (
+  waiverInterface: WaiverInterface,
+  action: WaiverReducerDispatch,
+) => {
+  const newWaiverInterface: WaiverInterface = { ...waiverInterface };
+  switch (action.type) {
+    case WaiverActions.LOADED_WAIVER: {
+      const optionalClauses: OptionalClauseResponse[] = [];
+      const requiredClauses: RequiredClauseResponse[] = [];
+      const { waiver } = action as LoadedWaiver;
+
+      waiver.clauses.forEach((clause: WaiverClause) => {
+        if (clause.required) requiredClauses.push(clause);
+        else optionalClauses.push({ ...clause, agreed: false });
+      });
+      newWaiverInterface.optionalClauses = optionalClauses;
+      newWaiverInterface.requiredClauses = requiredClauses;
+      newWaiverInterface.waiver = waiver;
+      newWaiverInterface.agreedRequiredClauses = false;
+      newWaiverInterface.loadingWaiver = false;
+      break;
+    }
+    case WaiverActions.ClICK_REQUIRED_CLAUSE: {
+      newWaiverInterface.agreedRequiredClauses = !newWaiverInterface.agreedRequiredClauses;
+      break;
+    }
+
+    case WaiverActions.CLICK_OPTIONAL_CLAUSE: {
+      if (!newWaiverInterface.optionalClauses) break;
+      const { optionalClauseId } = action as ClickOptionalClause;
+
+      const newOptionalClauses: OptionalClauseResponse[] = newWaiverInterface.optionalClauses?.map(
+        (
+          optionalClause: OptionalClauseResponse,
+          index: number,
+        ): OptionalClauseResponse => {
+          if (index === optionalClauseId) {
+            return { ...optionalClause, agreed: !optionalClause.agreed };
+          }
+          return optionalClause;
+        },
+      );
+      newWaiverInterface.optionalClauses = newOptionalClauses;
+      break;
+    }
+    case WaiverActions.WRITE_NAME: {
+      const { name } = action as FillName;
+      newWaiverInterface.wroteName = !!name;
+      break;
+    }
+    case WaiverActions.WRITE_DATE: {
+      const { date } = action as FillDate;
+      newWaiverInterface.wroteDate = !!date;
+      break;
+    }
+    default:
+  }
+  newWaiverInterface.waiverCompleted =
+    newWaiverInterface.agreedRequiredClauses &&
+    newWaiverInterface.wroteDate &&
+    newWaiverInterface.wroteName;
```

Comment:
Open for discussion - I'm not convinced this should be part of the waiver interface, and should be evaluated directly as a variable in `isWaiverFilled`. The waiver content doesn't _really_ need to know itself about whether it's completed, only the parent `Waiver` component

(it would also save the boilerplate that was added here vs. the previous iteration)

Reply:
Hmmm, my thought process was that we would want the ability for `WaiverInterface` to encapsulate everything to do with the waiver page (including the exit condition `waiverCompleted`). I wanted to include `waiverCompleted` and perform its logic inside `waiverReducer` since I felt that `RegistrantExperiencePage` mainly focused on the navigation between pages, only needing to know whether the waiverPage was completed and not the actual field of `WaiverInterface` to find out whether the waiver is complete. 

Though, I feel that this is just a personal style choice of mine and I see that both of our suggestions don't fundamentally change how the the codebase works.

Reply:
Yeah it's nothing blocking. Let's get this merged?

Reply:
Yep!